<a href="https://colab.research.google.com/github/IA2021UR/entregable-1-joferne/blob/main/3_comparacion_de_metodos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**1. IMPORTACIÓN DE LIBRERÍAS** 


In [40]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.linear_model import Perceptron
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint as sp_randint
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

**2. CARGA DE DATOS** 

In [2]:
!unzip ia2021.zip

Archive:  ia2021.zip
  inflating: test.csv                
  inflating: train.csv               


In [3]:
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')

**3. PREPROCESADO DE DATOS** 

Vamos a dividir el df_train en conjunto de entrenamiento (90%) y conjunto de validación (10%). 

In [4]:
X =  df_train.values[:,:-1]
Y =  df_train.values[:,-1]
(trainData, testData, trainLabels, testLabels) = train_test_split(X,Y,test_size=0.1, random_state=42)

Ahora normalizamos y estandarizamos los datos. En el análisis estadístico vimos que nos podría ser útil.

In [5]:
mms = MinMaxScaler()
trainData_norm = mms.fit_transform(trainData)
testData_norm = mms.transform(testData)

stdsc = StandardScaler()
trainData_std = stdsc.fit_transform(trainData)
testData_std = stdsc.transform(testData)

trainData_norm_std = stdsc.fit_transform(trainData_norm)
testData_norm_std = stdsc.transform(testData_norm)


**3. ALGORITMOS A UTILIZAR** 

In [6]:
def evalua(clasificador,testData,testLabels):
  num = 0
  for i in range(0,len(clasificador.predict(testData))):
    if clasificador.predict(testData)[i] == testLabels[i]:
      num = num + 1
  print('Cambian ' + str(len(testLabels) - num) + ' de ' + str(len(testLabels)))

def randomsearch(model,parameters):
    print("# Tuning hyper-parameters for %s" % score)
    print()

    clf = RandomizedSearchCV(model, parameters,cv=5)
    clf.fit(trainData, trainLabels)

    print("Best parameters set found on development set:")
    print()
    print(clf.best_params_)
    print()

def gridsearch(model,parameters,scores,k):
    for score in scores:
        print("# Tuning hyper-parameters for %s" % score)
        print()

        clf = GridSearchCV(model, parameters, cv=k,
                           scoring='%s_macro' % score)
        clf.fit(trainData, trainLabels)

        print("Best parameters set found on development set:")
        print()
        print(clf.best_params_)
        print()
        print("Grid scores on development set:")
        print()
        means = clf.cv_results_['mean_test_score']
        stds = clf.cv_results_['std_test_score']
        for mean, std, params in zip(means, stds, clf.cv_results_['params']):
            print("%0.3f (+/-%0.03f) for %r"
                  % (mean, std * 2, params))
        print()

**4. COMPARACIÓN DE MÉTODOS** 

Vamos a utilizar KNN, Árboles de decisión, SVM, Redes neuronales, Regresión logística y Regresión lineal múltiple. Primero haremos una pequeña aproximación y si vemos que tiene sentido haremos una búsqueda más exhaustiva de hiperparámetros.

KNN

In [7]:
modelKNN = KNeighborsClassifier(n_neighbors=5)
modelKNN.fit(trainData, trainLabels)
evalua(modelKNN,testData,testLabels)

Cambian 490 de 1200


Podría merecernos la pena hacer un ajuste de hiperparámetros.

In [8]:
for k in range(1,26,2):
    modelKNN = KNeighborsClassifier(n_neighbors=k)
    modelKNN.fit(trainData, trainLabels)
    score = modelKNN.score(testData,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=1, precisión=57.58%
k=3, precisión=58.42%
k=5, precisión=59.17%
k=7, precisión=59.50%
k=9, precisión=60.00%
k=11, precisión=59.58%
k=13, precisión=59.50%
k=15, precisión=59.67%
k=17, precisión=59.67%
k=19, precisión=59.17%
k=21, precisión=59.08%
k=23, precisión=58.92%
k=25, precisión=58.08%


In [9]:
for k in range(1,26,2):
    modelKNN = KNeighborsClassifier(n_neighbors=k)
    modelKNN.fit(trainData_norm, trainLabels)
    score = modelKNN.score(testData_norm,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=1, precisión=65.75%
k=3, precisión=64.92%
k=5, precisión=63.50%
k=7, precisión=63.17%
k=9, precisión=63.00%
k=11, precisión=63.00%
k=13, precisión=62.25%
k=15, precisión=61.67%
k=17, precisión=59.58%
k=19, precisión=60.67%
k=21, precisión=60.17%
k=23, precisión=59.83%
k=25, precisión=60.33%


In [10]:
for k in range(1,26,2):
    modelKNN = KNeighborsClassifier(n_neighbors=k)
    modelKNN.fit(trainData_std, trainLabels)
    score = modelKNN.score(testData_std,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=1, precisión=64.75%
k=3, precisión=63.42%
k=5, precisión=63.17%
k=7, precisión=62.67%
k=9, precisión=61.33%
k=11, precisión=61.75%
k=13, precisión=61.58%
k=15, precisión=60.08%
k=17, precisión=59.33%
k=19, precisión=59.42%
k=21, precisión=59.08%
k=23, precisión=58.33%
k=25, precisión=58.33%


In [11]:
for k in range(1,26,2):
    modelKNN = KNeighborsClassifier(n_neighbors=k)
    modelKNN.fit(trainData_norm_std, trainLabels)
    score = modelKNN.score(testData_norm_std,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=1, precisión=64.75%
k=3, precisión=63.42%
k=5, precisión=63.17%
k=7, precisión=62.67%
k=9, precisión=61.33%
k=11, precisión=61.75%
k=13, precisión=61.58%
k=15, precisión=60.08%
k=17, precisión=59.33%
k=19, precisión=59.42%
k=21, precisión=59.08%
k=23, precisión=58.33%
k=25, precisión=58.33%


In [12]:
modelKNN = KNeighborsClassifier(n_neighbors=1)
modelKNN.fit(trainData_norm, trainLabels)
evalua(modelKNN,testData_norm,testLabels)

Cambian 411 de 1200


In [13]:
modelKNN= KNeighborsClassifier()
randomsearch(modelKNN, {'weights': ['uniform','distance'], 
                        'p': [i for i in range(1,5)], 'n_neighbors': [j for j in range(1,30)]})

# Tuning hyper-parameters for 0.5833333333333334

Best parameters set found on development set:

{'weights': 'distance', 'p': 1, 'n_neighbors': 27}



In [14]:
modelKNN = KNeighborsClassifier(n_neighbors=13,weights='distance',p=1)
modelKNN.fit(trainData_norm, trainLabels)
evalua(modelKNN,testData_norm,testLabels)

Cambian 419 de 1200


In [15]:
gridsearch(modelKNN,[{'weights': ['uniform','distance'], 
                        'p': [i for i in range(1,5)], 'n_neighbors': [j for j in range(1,30)]}],['precision', 'recall'],5)

# Tuning hyper-parameters for precision

Best parameters set found on development set:

{'n_neighbors': 10, 'p': 1, 'weights': 'distance'}

Grid scores on development set:

0.606 (+/-0.020) for {'n_neighbors': 1, 'p': 1, 'weights': 'uniform'}
0.606 (+/-0.020) for {'n_neighbors': 1, 'p': 1, 'weights': 'distance'}
0.583 (+/-0.013) for {'n_neighbors': 1, 'p': 2, 'weights': 'uniform'}
0.583 (+/-0.013) for {'n_neighbors': 1, 'p': 2, 'weights': 'distance'}
0.580 (+/-0.016) for {'n_neighbors': 1, 'p': 3, 'weights': 'uniform'}
0.580 (+/-0.016) for {'n_neighbors': 1, 'p': 3, 'weights': 'distance'}
0.581 (+/-0.016) for {'n_neighbors': 1, 'p': 4, 'weights': 'uniform'}
0.581 (+/-0.016) for {'n_neighbors': 1, 'p': 4, 'weights': 'distance'}
0.617 (+/-0.020) for {'n_neighbors': 2, 'p': 1, 'weights': 'uniform'}
0.606 (+/-0.020) for {'n_neighbors': 2, 'p': 1, 'weights': 'distance'}
0.592 (+/-0.012) for {'n_neighbors': 2, 'p': 2, 'weights': 'uniform'}
0.583 (+/-0.013) for {'n_neighbors': 2, 'p': 2, 'wei

In [16]:
modelKNN = KNeighborsClassifier(n_neighbors=6,weights='distance',p=1)
modelKNN.fit(trainData_norm, trainLabels)
evalua(modelKNN,testData_norm,testLabels)

Cambian 409 de 1200


Árboles de decisión

In [17]:
modelArbol = DecisionTreeClassifier(random_state=84)
modelArbol.fit(trainData, trainLabels)
evalua(modelArbol,testData,testLabels)

Cambian 391 de 1200


Podría merecernos la pena hacer un mejor ajuste de hiperparámetros.

In [18]:
for k in range(2,11,1):
    modelArbol = DecisionTreeClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData, trainLabels)
    score = modelArbol.score(testData,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=67.42%
k=3, precisión=67.08%
k=4, precisión=66.42%
k=5, precisión=66.58%
k=6, precisión=66.25%
k=7, precisión=66.58%
k=8, precisión=66.50%
k=9, precisión=66.67%
k=10, precisión=66.58%


In [19]:
for k in range(2,11,1):
    modelArbol = DecisionTreeClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData_norm, trainLabels)
    score = modelArbol.score(testData_norm,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=66.67%
k=3, precisión=66.83%
k=4, precisión=66.50%
k=5, precisión=66.42%
k=6, precisión=66.50%
k=7, precisión=67.00%
k=8, precisión=66.42%
k=9, precisión=66.75%
k=10, precisión=66.75%


In [20]:
for k in range(2,11,1):
    modelArbol = DecisionTreeClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData_std, trainLabels)
    score = modelArbol.score(testData_std,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=67.25%
k=3, precisión=66.42%
k=4, precisión=66.58%
k=5, precisión=66.25%
k=6, precisión=66.17%
k=7, precisión=66.67%
k=8, precisión=66.50%
k=9, precisión=66.42%
k=10, precisión=66.58%


In [21]:
for k in range(2,11,1):
    modelArbol = DecisionTreeClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData_norm_std, trainLabels)
    score = modelArbol.score(testData_norm_std,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=67.25%
k=3, precisión=66.42%
k=4, precisión=66.58%
k=5, precisión=66.25%
k=6, precisión=66.17%
k=7, precisión=66.67%
k=8, precisión=66.50%
k=9, precisión=66.42%
k=10, precisión=66.58%


In [22]:
for k in range(2,11,1):
    modelArbol = DecisionTreeClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData_norm_std, trainLabels)
    score = modelArbol.score(testData_norm_std,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=67.25%
k=3, precisión=66.42%
k=4, precisión=66.58%
k=5, precisión=66.25%
k=6, precisión=66.17%
k=7, precisión=66.67%
k=8, precisión=66.50%
k=9, precisión=66.42%
k=10, precisión=66.58%


In [23]:
modelArbol = DecisionTreeClassifier(min_samples_split=2,random_state=84)
modelArbol.fit(trainData_norm_std, trainLabels)
evalua(modelArbol,testData_norm_std,testLabels)

Cambian 393 de 1200


In [24]:
modelArbol = DecisionTreeClassifier()
randomsearch(modelArbol, {'criterion': ['gini','entropy'], 
                        'min_samples_split': [i for i in range(2,11,1)], 'max_features': ['auto','sqrt','log2']})

# Tuning hyper-parameters for 0.6658333333333334

Best parameters set found on development set:

{'min_samples_split': 8, 'max_features': 'auto', 'criterion': 'gini'}



In [25]:
modelArbol = DecisionTreeClassifier(criterion='entropy',min_samples_split=2,random_state=84)
modelArbol.fit(trainData_norm_std, trainLabels)
evalua(modelArbol,testData_norm_std,testLabels)

Cambian 382 de 1200


In [26]:
gridsearch(modelArbol,[{'criterion': ['gini','entropy'], 
                        'min_samples_split': [i for i in range(2,11,1)], 'max_features': ['auto','sqrt','log2']}],['precision', 'recall'],5)

# Tuning hyper-parameters for precision

Best parameters set found on development set:

{'criterion': 'entropy', 'max_features': 'auto', 'min_samples_split': 2}

Grid scores on development set:

0.661 (+/-0.020) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 2}
0.651 (+/-0.018) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 3}
0.657 (+/-0.018) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 4}
0.660 (+/-0.008) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 5}
0.666 (+/-0.012) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 6}
0.660 (+/-0.013) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 7}
0.656 (+/-0.027) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 8}
0.652 (+/-0.018) for {'criterion': 'gini', 'max_features': 'auto', 'min_samples_split': 9}
0.662 (+/-0.011) for {'criterion': 'gini', 'max_features': 'auto', 'min_sampl

In [27]:
for k in range(2,11,1):
    modelArbol = DecisionTreeClassifier(criterion='entropy',min_samples_split=k,max_features='log2',random_state=84)
    scores = cross_val_score(modelArbol,trainData_norm_std,trainLabels,cv=10)
    print("k=%d, Precisión: %0.2f (+/- %0.2f)" % (k, scores.mean(), scores.std() * 2))

k=2, Precisión: 0.68 (+/- 0.02)
k=3, Precisión: 0.68 (+/- 0.03)
k=4, Precisión: 0.68 (+/- 0.03)
k=5, Precisión: 0.69 (+/- 0.02)
k=6, Precisión: 0.68 (+/- 0.03)
k=7, Precisión: 0.68 (+/- 0.02)
k=8, Precisión: 0.68 (+/- 0.03)
k=9, Precisión: 0.68 (+/- 0.02)
k=10, Precisión: 0.68 (+/- 0.03)


In [28]:
modelArbol = DecisionTreeClassifier(criterion='entropy',min_samples_split=5,max_features='log2',random_state=84)
modelArbol.fit(trainData_norm_std, trainLabels)
evalua(modelArbol,testData_norm_std,testLabels)

Cambian 373 de 1200


Regresión logística

In [29]:
modelLR = LogisticRegression(random_state=84)
modelLR.fit(trainData, trainLabels)
evalua(modelLR,testData,testLabels)

Cambian 829 de 1200


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


Este método se ve especialmente inapropiado para este conjunto de datos. No lo vamos a seguir estudiando.



SVM

In [32]:
'''modelSVMLineal = SVC(kernel="linear")
modelSVMLineal.fit(trainData, trainLabels)
evalua(modelSVMLineal,testData,testLabels)'''

# No consigue ejecutarlo

'modelSVMLineal = SVC(kernel="linear")\nmodelSVMLineal.fit(trainData, trainLabels)\nevalua(modelSVMLineal,testData,testLabels)'

In [33]:
modelSVMSigmoide = SVC(kernel="sigmoid",gamma=2,coef0=1)
modelSVMSigmoide.fit(trainData, trainLabels)
evalua(modelSVMSigmoide,testData,testLabels)

Cambian 850 de 1200


Este método se ve especialmente inapropiado para este conjunto de datos. No lo vamos a seguir estudiando.

In [34]:
modelPerceptron = Perceptron(max_iter=10,eta0=1.0,random_state=84)
modelPerceptron.fit(trainData,trainLabels)
evalua(modelPerceptron,testData,testLabels)

Cambian 950 de 1200


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_stochastic_gradient.py:557: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)


In [35]:
modelMultiLayerPerceptron = MLPClassifier(hidden_layer_sizes=(5,2),learning_rate_init=0.001,momentum=0.9)
modelMultiLayerPerceptron.fit(trainData,trainLabels)
evalua(modelMultiLayerPerceptron,testData,testLabels)


Cambian 853 de 1200


Este método se ve especialmente inapropiado para este conjunto de datos. No lo vamos a seguir estudiando.

Regresión lineal

Hemos visto que la correlación entre las variables era casi nula, por tanto debería ser un modelo que dé malos resultados.

In [36]:
slr = LinearRegression(normalize=True)
slr.fit(trainData_std,trainLabels)
y_train_pred = slr.predict(trainData_std)
y_test_pred = slr.predict(testData_std)

In [37]:
print('Valor MSE entrenamiento: %.3f, valor MSE test: %.3f' % 
      (mean_squared_error(trainLabels,y_train_pred),
      mean_squared_error(testLabels,y_test_pred)))

Valor MSE entrenamiento: 2.899, valor MSE test: 3.047


In [38]:
y_test_pred_round = []
for elem in y_test_pred:
  y_test_pred_round.append(round(elem))
y_test_pred_round = np.array(y_test_pred_round)

In [39]:
num = 0
for i in range(0,len(y_test_pred_round)):
  if y_test_pred_round[i] == testLabels[i]:
    num = num + 1
print('Cambian ' + str(len(testLabels) - num) + ' de ' + str(len(testLabels)))

Cambian 1044 de 1200


RandomForest

En vista de que los árboles de decisión son el modelo que mejor se ha ajustado vamos a probar con los RandomForest.

In [41]:
forest = RandomForestClassifier(random_state=84)
forest.fit(trainData, trainLabels)
evalua(forest,testData,testLabels)

Cambian 313 de 1200


Este hasta el momento es el método que mejor se ha ajustado. Sin embargo por cómo está implementado sabemos que la búsqueda de los hiperparámetros adecuados no mejorará mucho el modelo. No obstante vamos a tratar de ajustarlo lo mejor posible.

In [42]:
for k in range(2,15,1):
    modelArbol = RandomForestClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData, trainLabels)
    score = modelArbol.score(testData,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=73.92%
k=3, precisión=73.50%
k=4, precisión=73.75%
k=5, precisión=74.08%
k=6, precisión=73.92%
k=7, precisión=74.25%
k=8, precisión=73.50%
k=9, precisión=73.50%
k=10, precisión=74.08%
k=11, precisión=74.17%
k=12, precisión=73.67%
k=13, precisión=73.50%
k=14, precisión=73.42%


In [43]:
for k in range(2,15,1):
    modelArbol = RandomForestClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData_std, trainLabels)
    score = modelArbol.score(testData_std,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=73.58%
k=3, precisión=73.50%
k=4, precisión=73.83%
k=5, precisión=74.08%
k=6, precisión=74.17%
k=7, precisión=73.92%
k=8, precisión=73.83%
k=9, precisión=73.83%
k=10, precisión=73.50%
k=11, precisión=74.08%
k=12, precisión=73.58%
k=13, precisión=73.17%
k=14, precisión=73.58%


In [44]:
for k in range(2,15,1):
    modelArbol = RandomForestClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData_norm, trainLabels)
    score = modelArbol.score(testData_norm,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=73.67%
k=3, precisión=73.83%
k=4, precisión=74.67%
k=5, precisión=74.50%
k=6, precisión=74.17%
k=7, precisión=74.08%
k=8, precisión=73.75%
k=9, precisión=73.83%
k=10, precisión=74.08%
k=11, precisión=73.67%
k=12, precisión=73.50%
k=13, precisión=73.92%
k=14, precisión=73.75%


In [45]:
for k in range(2,15,1):
    modelArbol = RandomForestClassifier(min_samples_split=k,random_state=84)
    modelArbol.fit(trainData_norm_std, trainLabels)
    score = modelArbol.score(testData_norm_std,testLabels)
    print("k=%d, precisión=%.2f%%" % (k, score * 100))

k=2, precisión=73.58%
k=3, precisión=73.50%
k=4, precisión=73.83%
k=5, precisión=74.08%
k=6, precisión=74.17%
k=7, precisión=73.92%
k=8, precisión=73.83%
k=9, precisión=73.83%
k=10, precisión=73.50%
k=11, precisión=74.08%
k=12, precisión=73.58%
k=13, precisión=73.17%
k=14, precisión=73.58%


In [49]:
forest = RandomForestClassifier(min_samples_split=4,random_state=84)
forest.fit(trainData_norm, trainLabels)
evalua(forest,testData_norm,testLabels)

Cambian 304 de 1200


In [47]:
forest = RandomForestClassifier()
randomsearch(modelArbol, {'criterion': ['gini','entropy'], 
                        'min_samples_split': [i for i in range(2,11,1)], 'max_features': ['auto','sqrt','log2']})

# Tuning hyper-parameters for 0.7358333333333333

Best parameters set found on development set:

{'min_samples_split': 3, 'max_features': 'auto', 'criterion': 'entropy'}



In [50]:
forest = RandomForestClassifier(min_samples_split=3,criterion='entropy',random_state=84)
forest.fit(trainData_norm, trainLabels)
evalua(forest,testData_norm,testLabels)

Cambian 312 de 1200


**5. CONCLUSIONES** 

En el mejor modelo que hemos encontrado para los métodos en los cuáles no hemos ajustado hiperparámetros:
- Regresión logística: cambian 829 de 1200 (30.92% acierto)
- SVM: cambian 850 de 1200 (29.12% acierto)
- Redes neuronales: cambian 850 de 1200 (29.12% acierto)
- Regresión lineal: cambian 1044 de 1200 (23.0% acierto)

Para los que sí hemos hecho una búsqueda más exhaustiva:
- KNN: cambian 409 de 1200 (65.92% de acierto)
- Árboles de decisión: cambian 373 de 1200 (68.92% acierto)
- RandomForest: cambian 304 de 1200 (74.67% acierto)
